sync/threadgroup.go

```diff
@@ -14,11 +15,18 @@ var ErrStopped = errors.New("ThreadGroup already stopped")
 // notifying callers when shutdown occurrs. Another key difference is that a
 // ThreadGroup is only intended be used once; as such, its Add and Stop
 // methods return errors if Stop has already been called.
+//
+// During shutdown, it is common to close resources such as net.Conns and
+// net.Listeners. Typically, this would require spawning a goroutine to wait
+// on the ThreadGroup's StopChan and then close the resource. To make this
+// more convenient, ThreadGroup provides a RegisterCloser method. io.Closers
+// registered in this manner will be automatically closed when Stop is called.
 type ThreadGroup struct {
 	stopChan chan struct{}
 	chanOnce sync.Once
 	mu       sync.Mutex
 	wg       sync.WaitGroup
+	closers  []io.Closer
 }
```

Comment:
Just thinking about memory and garbage collection... I'm not sure that there's an easy way to tell when an object is garbage collected, but if you are holding a slice of `closers` it means that none of those objects will be GC'd until the `tg` is GC'd, even if the `tg` is already closed.

If everything is running smoothly, the `tg` will be GC'd shortly after it is shutdown, but we may want to wipe the list of closers anyway, to minimize the damage in the event of memory leaks.